dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/event/JdbcEventStore.java

```diff
@@ -603,141 +600,113 @@ lateral jsonb_each(
-          .append(SPACE)
-          .append(":" + parameterKey);
+
+      StringBuilder filterString = new StringBuilder();
+      filterString.append(queryCol).append(SPACE);
+
+      filterString.append(
+          switch (filter.getOperator()) {
+            case NULL, NNULL -> filter.getSqlOperator() + SPACE;
+            default -> {
+              mapSqlParameterSource.addValue(
+                  parameterKey,
+                  isNumericTea
+                      ? Double.valueOf(filter.getSqlBindFilter())
+                      : StringUtils.lowerCase(filter.getSqlBindFilter()),
+                  itemType);
+              yield new StringBuilder()
+                  .append(filter.getSqlOperator())
+                  .append(SPACE)
+                  .append(":")
+                  .append(parameterKey)
+                  .append(SPACE);
+            }
+          });
+
       filterStrings.add(filterString.toString());
     }
     query.append(String.join(AND, filterStrings));
 
-    if (isNumericTea) {
-      query.append(" END ");
-    }
-
     return query.toString();
   }
 
   /**
-   * Generates the LEFT JOINs used for attributes we are ordering by (If any). We use LEFT JOIN to
-   * avoid removing any rows if there is no value for a given attribute and te. The result of this
-   * LEFT JOIN is used in the sub-query projection, and ordering in the sub-query and main query.
+   * Generates the LEFT JOIN based on the attributes we are ordering and filtering by, if any. We
+   * use LEFT JOIN to avoid removing any rows if there is no value for a given attribute and te. The
+   * result of this LEFT JOIN is used in the sub-query projection, and ordering in the sub-query and
+   * main query.
+   *
+   * <p>Attribute filtering is handled in {@link
+   * #getWhereClauseFromAttributeFilterConditions(EventQueryParams, MapSqlParameterSource,
+   * SqlHelper)}.
    *
-   * @return a SQL LEFT JOIN for attributes used for ordering, or empty string if no attributes is
-   *     used in order.
+   * @return a SQL LEFT JOIN for the relevant attributes, or an empty string if none are provided.
    */
-  private String getFromSubQueryJoinOrderByAttributes(EventQueryParams params) {
-    StringBuilder joinOrderAttributes = new StringBuilder();
+  private String getLeftJoinFromAttributes(EventQueryParams params) {
+    StringBuilder attributes = new StringBuilder();
 
-    for (TrackedEntityAttribute orderAttribute : params.leftJoinAttributes()) {
-
-      joinOrderAttributes
+    for (TrackedEntityAttribute attribute : params.leftJoinAttributes()) {
+      attributes
           .append(" left join trackedentityattributevalue as ")
-          .append(quote(orderAttribute.getUid()))
+          .append(quote(attribute.getUid()))
           .append(" on ")
-          .append(quote(orderAttribute.getUid()))
+          .append(quote(attribute.getUid()))
           .append(".trackedentityid = TE.trackedentityid ")
           .append("and ")
```

Comment:
It might be a drop in the ocean but I don't get why we make calls to append for strings we should concatenate right away like `".trackedentityid = TE.trackedentityid and "`.